internal/engine/wazevo/engine.go

```diff
@@ -0,0 +1,206 @@
+
+	importedFns, localFns := int(module.ImportFunctionCount), len(module.FunctionSection)
+	if importedFns+localFns == 0 {
+		e.addCompiledModule(module, cm)
+		return nil
+	}
+
+	// TODO: reuse the map to avoid allocation.
+	exportedFnIndex := make(map[wasm.Index]struct{}, len(module.ExportSection))
+	for i := range module.ExportSection {
+		if module.ExportSection[i].Type == wasm.ExternTypeFunc {
+			exportedFnIndex[module.ExportSection[i].Index] = struct{}{}
+		}
+	}
+
+	// Creates new compiler instances which are reused for each function.
+	ssaBuilder := ssa.NewBuilder()
+	fe := frontend.NewFrontendCompiler(module, ssaBuilder)
+	machine := newMachine()
+	be := backend.NewCompiler(machine, ssaBuilder)
+
+	totalSize := 0 // Total binary size of the executable.
+	cm.functionsOffsets = make([]int, localFns)
+	bodies := make([][]byte, localFns)
+	for i := range module.CodeSection {
+		fidx := wasm.Index(i + importedFns)
+		fref := frontend.FunctionIndexToFuncRef(fidx)
+
+		_, needGoEntryPreamble := exportedFnIndex[fidx]
+
+		// Align 16-bytes boundary.
+		totalSize = (totalSize + 15) &^ 15
+		cm.functionsOffsets[i] = totalSize
+
+		typ := &module.TypeSection[module.FunctionSection[i]]
+
+		codeSeg := &module.CodeSection[i]
+		if codeSeg.GoFunc != nil {
+			panic("TODO: host module")
+		}
+
+		// Initializes both frontend and backend compilers.
+		fe.Init(wasm.Index(i), typ, codeSeg.LocalTypes, codeSeg.Body)
+		be.Init(needGoEntryPreamble)
+
+		// Lower Wasm to SSA.
+		err := fe.LowerToSSA()
+		if err != nil {
+			return fmt.Errorf("wasm->ssa: %v", err)
+		}
+
+		// Run SSA-level optimization passes.
+		ssaBuilder.RunPasses()
+
+		// Finalize the layout of SSA blocks which might use the optimization results.
+		ssaBuilder.LayoutBlocks()
+
+		// Now our ssaBuilder contains the necessary information to further lower them to
+		// machine code.
+		body, rels, goPreambleSize, err := be.Compile()
+		if err != nil {
+			return fmt.Errorf("ssa->machine code: %v", err)
+		}
+
+		e.refToBinaryOffset[fref] = totalSize +
+			// During the relocation, call target needs to be the beginning of function after Go entry preamble.
+			goPreambleSize
+
+		// At this point, relocation offsets are relative to the start of the function body,
+		// so we adjust it to the start of the executable.
+		for _, r := range rels {
+			r.Offset += int64(totalSize)
+			e.rels = append(e.rels, r)
+		}
+
+		// TODO: optimize as zero copy.
+		copied := make([]byte, len(body))
+		copy(copied, body)
+		bodies[i] = copied
+		totalSize += len(body)
+	}
+
+	// Allocate executable memory and then copy the generated machine code.
+	executable, err := platform.MmapCodeSegment(totalSize)
+	if err != nil {
+		panic(err)
+	}
+	cm.executable = executable
+
+	for i, b := range bodies {
+		offset := cm.functionsOffsets[i]
+		copy(executable[offset:], b)
+	}
+
+	// Resolve relocations for local function calls.
+	machine.ResolveRelocations(e.refToBinaryOffset, executable, e.rels)
```

Comment:
after compiling all local functions, we can resolve the relocations (relative offers between the call instructions and the target functions).